rosbag2_compression/src/rosbag2_compression/zstd_compressor.cpp

```diff
@@ -0,0 +1,147 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <cstdio>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "rosbag2_compression/zstd_compressor.hpp"
+
+#include "rosbag2_storage/filesystem_helper.hpp"
+
+#include "logging.hpp"
+
+namespace
+{
+// Increasing the compression level will:
+//   - Increase the time taken to compress
+//   - Decrease the size of the compressed data
+// Setting to zero uses Zstd's default value of 3.
+constexpr const int DEFAULT_ZSTD_COMPRESSION_LEVEL = 1;
+
+constexpr const char COMPRESSION_IDENTIFIER[] = "zstd";
+
+FILE * open_file(const std::string & uri, const std::string & read_mode)
+{
+  FILE * fp{nullptr};
+#ifdef _WIN32
+  fopen_s(&fp, uri.c_str(), read_mode.c_str());
+#else
+  fp = std::fopen(uri.c_str(), read_mode.c_str());
+#endif
+  return fp;
+}
+
+std::vector<uint8_t> get_input_buffer(
+  const std::string & uri,
+  size_t & decompressed_buffer_length)
+{
+  // Get the file size
+  decompressed_buffer_length = rosbag2_storage::FilesystemHelper::get_file_size(uri);
+  // Read in buffer, handling accordingly
+  auto file_pointer = open_file(uri.c_str(), "rb");
+  if (file_pointer == nullptr) {
+    throw std::runtime_error("Error opening file");
+  }
+  // Allocate and read in
+  std::vector<uint8_t> decompressed_buffer;
+  decompressed_buffer.reserve(decompressed_buffer_length);
+  fread(decompressed_buffer.data(), sizeof(uint8_t), decompressed_buffer_length, file_pointer);
```

Comment:
This line results in a compiler warning: https://ci.ros2.org/job/ci_packaging_linux/279/warnings23Result/new/

Please resolve it as soon as feasible.

Reply:
See #237.